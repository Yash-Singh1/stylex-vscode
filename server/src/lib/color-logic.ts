// Color logic from Tailwind CSS IntelliSense

import * as culori from "culori";
import * as namedColors from "color-name";
import type { Color } from "vscode-languageserver";
import { StringLiteral } from "@swc/wasm-web";

export type KeywordColor = "transparent" | "currentColor";

export function getColorFromValue(
  value: unknown,
): culori.Color | KeywordColor | null {
  // console.log(typeof value);

  if (typeof value !== "string") return null;

  const trimmedValue = value.trim();

  if (trimmedValue.toLowerCase() === "transparent") {
    return "transparent";
  }

  if (trimmedValue.toLowerCase() === "currentcolor") {
    return "currentColor";
  }

  if (
    !/^\s*(?:rgba?|hsla?)\s*\([^)]+\)\s*$/.test(trimmedValue) &&
    !/^\s*#[0-9a-f]+\s*$/i.test(trimmedValue) &&
    !Object.keys(
      (namedColors as unknown as { default: typeof import("color-name") })
        .default,
    ).includes(trimmedValue)
  ) {
    // console.log('fail here', trimmedValue, Object.keys(namedColors));
    return null;
  }

  const color = culori.parse(trimmedValue);
  // console.log('unparsable', color);
  return color ?? null;
}

const toRgb = culori.converter("rgb");

export function culoriColorToVscodeColor(color: culori.Color): Color {
  const rgb = toRgb(color);
  return { red: rgb.r, green: rgb.g, blue: rgb.b, alpha: rgb.alpha ?? 1 };
}
