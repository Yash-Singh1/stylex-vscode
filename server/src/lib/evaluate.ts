import type {
	ArrayExpression,
	ComputedPropName,
	Expression,
	Span,
	TemplateElement,
} from '@swc/wasm-web';
import StateManager from './state-manager';

type ResultType =
	| /* prettier-ignore */ {
			value: InnerType;
			static: false;
		}
	| /* prettier-ignore */ {
			value: InnerType;
			static: true;
			span: Span;
		}
	| /* prettier-ignore */ {
			id: string;
			static: true;
			span: Span;
		};

type InnerType =
	| ResultType[]
	| string
	| number
	| boolean
	| null
	| undefined
	| RegExp
	| Record<string, ResultType>
	| bigint;

function processArrayExpression(node: ArrayExpression, stateManager: StateManager) {
	const result = node.elements.reduce<any[]>((accumalator, exprOrSpread) => {
		if (!exprOrSpread || !accumalator) return accumalator;
		if (exprOrSpread.spread) {
			const result = evaluate(exprOrSpread.expression, stateManager);

			if (!result.static) {
				accumalator.push(result);
			}

			accumalator.push(
				...(<any[]>(
					(result &&
					typeof result === 'object' &&
					(Array.isArray(result) ||
						(Symbol.iterator in result && typeof result[Symbol.iterator] === 'function'))
						? result
						: [result])
				))
			);
		} else {
			accumalator.push(evaluate(exprOrSpread.expression, stateManager));
		}
		return accumalator;
	}, []);

	return { value: result, static: true, span: node.span } satisfies ResultType;
}

export function evaluate(
	node: Expression | ComputedPropName | TemplateElement,
	stateManager: StateManager
): ResultType {
	switch (node.type) {
		case 'StringLiteral':
		case 'NumericLiteral':
		case 'BooleanLiteral':
		case 'BigIntLiteral':
			return { value: node.value, static: true, span: node.span };
		case 'NullLiteral':
			return { value: null, static: true, span: node.span };
		case 'RegExpLiteral':
			return {
				value: new RegExp(node.pattern, node.flags),
				static: true,
				span: node.span,
			};
		case 'Invalid':
			throw new Error('Invalid expression');
		case 'Identifier':
			if (node.value === 'undefined')
				return { value: undefined, static: true, span: node.span };
			else {
				return { id: node.value, static: true, span: node.span };
				// TODO: Expose scope manager
				throw new Error('Identifier not implemented');
			}
		case 'ArrayExpression': {
			return processArrayExpression(node, stateManager);
		}

		case 'ObjectExpression': {
			const result = node.properties.reduce<Record<string, ResultType>>(
				(accumalator, property) => {
					if (property.type === 'SpreadElement') {
						const result = evaluate(property.arguments, stateManager);
						if (result && typeof result === 'object' && 'value' in result) {
							if (!result.static) return accumalator;
							accumalator = Object.assign(accumalator, result.value);
						}
						return accumalator;
					}

					switch (property.type) {
						case 'KeyValueProperty':
						case 'AssignmentProperty': {
							const keyVal = evaluate(property.key, stateManager);
							if (
								typeof keyVal === 'string' ||
								typeof keyVal === 'number' ||
								typeof keyVal === 'symbol'
							) {
								accumalator[keyVal] = evaluate(property.value, stateManager);
							}
							break;
						}

						case 'GetterProperty':
						case 'SetterProperty':
						case 'MethodProperty': {
							return accumalator;
						}

						case 'Identifier': {
							const result = evaluate(property, stateManager);
							if (!result.static) return accumalator;
							accumalator[property.value] = result;
							break;
						}
					}

					return accumalator;
				},
				{}
			);

			return { value: result, static: true, span: node.span };
		}

		case 'ArrowFunctionExpression':
		case 'FunctionExpression':
			return { value: undefined, static: false };

		case 'AwaitExpression': {
			const result = evaluate(node.argument, stateManager);
			if (!result.static) return { value: undefined, static: false };
			return result;
		}

		case 'BinaryExpression': {
			const left = evaluate(node.left, stateManager);
			const right = evaluate(node.right, stateManager);

			if (!left.static || !right.static) return { value: undefined, static: false };
			let result;

			if ('id' in left) return { value: undefined, static: false };
			if ('id' in right) return { value: undefined, static: false };

			switch (node.operator) {
				case '==':
					result = left.value == right.value;
					break;
				case '!=':
					result = left.value != right.value;
					break;
				case '===':
					result = left.value === right.value;
					break;
				case '!==':
					result = left.value !== right.value;
					break;
				case '<':
					if (left.value == null || right.value == null) {
						result = false;
					} else result = left.value < right.value;
					break;
				case '<=':
					if (left.value == null || right.value == null) {
						result = false;
					} else result = left.value <= right.value;
					break;
				case '>':
					if (left.value == null || right.value == null) {
						result = false;
					} else result = left.value > right.value;
					break;
				case '>=':
					if (left.value == null || right.value == null) {
						result = false;
					} else result = left.value >= right.value;
					break;
				case 'in':
					try {
						// @ts-expect-error -- Error handled
						result = left.value in right.value;
					} catch {
						result = false;
					}
					break;
				case 'instanceof':
					try {
						// @ts-expect-error -- Error handled
						result = left.value instanceof right.value;
					} catch {
						result = false;
					}
					break;
				case '&&':
					result = left.value && right.value;
					break;
				case '||':
					result = left.value || right.value;
					break;
				case '??':
					result = left.value ?? right.value;
					break;
			}
			if (typeof result !== 'undefined')
				return { value: result, static: true, span: node.span };

			if (
				typeof left.value !== 'number' ||
				typeof left.value !== 'bigint' ||
				typeof right.value !== 'number' ||
				typeof right.value !== 'bigint' ||
				typeof left.value !== typeof right.value
			) {
				return { value: undefined, static: true, span: node.span };
			}

			switch (node.operator) {
				case '<<':
					result = left.value << right.value;
					break;
				case '>>':
					result = left.value >> right.value;
					break;
				case '>>>':
					result = left.value >>> right.value;
					break;
				case '+':
					result = left.value + right.value;
					break;
				case '-':
					result = left.value - right.value;
					break;
				case '*':
					result = left.value * right.value;
					break;
				case '/':
					result = left.value / right.value;
					break;
				case '%':
					result = left.value % right.value;
					break;
				case '**':
					result = left.value ** right.value;
					break;
				case '|':
					result = left.value | right.value;
					break;
				case '^':
					result = left.value ^ right.value;
					break;
				case '&':
					result = left.value & right.value;
					break;
				default:
					throw new Error('Unknown binary operator');
			}

			break;
		}

		case 'JSXElement':
		case 'JSXFragment':
		case 'JSXEmptyExpression':
		case 'JSXMemberExpression':
		case 'JSXNamespacedName':
		case 'JSXText':
			return { value: undefined, static: false };

		case 'TsAsExpression':
		case 'TsNonNullExpression':
		case 'TsConstAssertion':
		case 'TsTypeAssertion':
		case 'TsInstantiation':
			return evaluate(node.expression, stateManager);

		case 'ParenthesisExpression':
			return evaluate(node.expression, stateManager);

		case 'UnaryExpression': {
			const result = evaluate(node.argument, stateManager);
			if (!result.static || 'id' in result) return { value: undefined, static: false };
			if (result.value == null) return { value: undefined, static: true, span: node.span };

			switch (node.operator) {
				case '+':
					try {
						// @ts-expect-error -- Error handled
						return { value: +result.value, static: true, span: node.span };
					} catch {
						return { value: undefined, static: true, span: node.span };
					}
				case '-':
					return { value: -result.value, static: true, span: node.span };
				case '!':
					return { value: !result.value, static: true, span: node.span };
				case '~':
					return { value: ~result.value, static: true, span: node.span };
				case 'typeof':
					return { value: typeof result.value, static: true, span: node.span };
				case 'void':
					return { value: undefined, static: true, span: node.span };
				case 'delete':
					return { value: undefined, static: false };
				default:
					throw new Error('Unknown unary operator');
			}
		}

		case 'AssignmentExpression':
			return evaluate(node.right, stateManager);

		case 'ThisExpression':
			return { value: undefined, static: false };

		case 'ConditionalExpression': {
			const test = evaluate(node.test, stateManager);
			if (!test.static || 'id' in test) return { value: undefined, static: false };

			if (test.value) {
				return evaluate(node.consequent, stateManager);
			} else {
				return evaluate(node.alternate, stateManager);
			}
		}

		case 'SuperPropExpression':
			return { value: undefined, static: false };

		case 'TaggedTemplateExpression':
			return { value: undefined, static: false };

		case 'TemplateLiteral': {
			const values = [];
			let quasisIndex = 0;
			let expressionsIndex = 0;
			let currentQuasi = true;
			while (
				currentQuasi
					? quasisIndex < node.quasis.length
					: expressionsIndex < node.expressions.length
			) {
				if (currentQuasi) {
					const result = evaluate(node.quasis[quasisIndex], stateManager);
					if (!result.static || 'id' in result) return { value: undefined, static: false };
					values.push(result.value);
					++quasisIndex;
				} else {
					const result = evaluate(node.expressions[expressionsIndex], stateManager);
					if (!result.static || 'id' in result) return { value: undefined, static: false };
					values.push(result.value);
					++expressionsIndex;
				}
				currentQuasi = !currentQuasi;
			}

			return { value: values.join(''), static: true, span: node.span };
		}

		case 'TemplateElement':
			return { value: node.raw, static: true, span: node.span };

		case 'NewExpression':
			return { value: undefined, static: false };

		case 'ClassExpression':
			return { value: undefined, static: false };

		case 'CallExpression': {
			if (
				(node.callee.type === 'MemberExpression' &&
					node.callee.object.type === 'Identifier' &&
					stateManager.verifyStylexIdentifier(node.callee.object.value) &&
					node.callee.property.type === 'Identifier' &&
					node.callee.property.value === 'firstThatWorks') ||
				(node.callee.type === 'Identifier' &&
					stateManager.verifyNamedImport(node.callee.value) === 'firstThatWorks')
			) {
				const result = processArrayExpression(
					{
						type: 'ArrayExpression',
						span: node.span,
						elements: node.arguments,
					},
					stateManager
				);

				result.value = result.value.reverse();

				return result;
			}

			return { value: undefined, static: false };
		}

		case 'Computed': {
			return evaluate(node.expression, stateManager);
		}

		case 'MemberExpression': {
			const object = evaluate(node.object, stateManager);
			if (!object.static) return { value: undefined, static: false };
			const property = evaluate(node.property, stateManager);
			if (!property.static) return { value: undefined, static: false };

			const propertyKey = 'id' in property ? property.id : property.value;

			return 'id' in object
				? /* prettier-ignore */ {
						id: `${object.id}.${propertyKey}`,
						static: true,
						span: node.span,
					}
				: /* prettier-ignore */ {
						value:
							object.value == null
								? object.value
								: /* prettier-ignore */ typeof propertyKey === 'string'
						? // @ts-expect-error -- Ignore member expression strict rules
							object.value[propertyKey]
						: undefined,
						static: true,
						span: node.span,
					};
		}

		case 'OptionalChainingExpression':
			return { value: undefined, static: false };

		case 'UpdateExpression':
			return { value: undefined, static: false };

		case 'YieldExpression':
			return { value: undefined, static: false };

		case 'MetaProperty':
			return { value: undefined, static: false };

		case 'PrivateName':
			return evaluate(node.id, stateManager);

		case 'SequenceExpression':
			return evaluate(node.expressions[node.expressions.length - 1], stateManager);

		default:
			throw new Error('Unknown expression');
	}

	return { value: undefined, static: false };
}
