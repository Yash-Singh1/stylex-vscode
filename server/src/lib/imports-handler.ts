import type { ImportDeclaration, VariableDeclarator } from "@swc/types";
import StateManager from "./state-manager";

export function handleImports(node: ImportDeclaration, state: StateManager) {
  // TODO: Configuration option for stylex package name or scan for it
  if (node.source.value !== "@stylexjs/stylex") {
    return;
  }

  for (const specifier of node.specifiers) {
    switch (specifier.type) {
      case "ImportDefaultSpecifier":
        state.addStylexIdentifier(specifier.local.value);
        break;

      case "ImportSpecifier": {
        if (specifier.imported && specifier.imported.value === "default") {
          state.addStylexIdentifier(specifier.local.value);
        } else if (specifier.imported) {
          state.addNamedImport(specifier.local.value, specifier.imported.value);
        } else {
          state.addNamedImport(specifier.local.value, specifier.local.value);
        }
        break;
      }

      case "ImportNamespaceSpecifier":
        state.addStylexIdentifier(specifier.local.value);
        break;
    }
  }
}

export function handleRequires(node: VariableDeclarator, state: StateManager) {
  if (
    node.init?.type !== "CallExpression" ||
    node.init.callee.type !== "Identifier" ||
    node.init.callee.value !== "require" ||
    node.init.arguments[0].expression.type !== "StringLiteral"
  ) {
    return;
  }

  if (node.init.arguments[0].expression.value !== "@stylexjs/stylex") {
    return;
  }

  switch (node.id.type) {
    case "Identifier":
      state.addStylexIdentifier(node.id.value);
      break;

    case "ObjectPattern":
      for (const property of node.id.properties) {
        if (property.type === "AssignmentPatternProperty") {
          state.addNamedImport(property.key.value, property.key.value);
        } else if (property.type === "KeyValuePatternProperty") {
          // TODO: Handle computed property keys
          if (
            property.value.type === "Identifier" &&
            property.key.type !== "Computed"
          ) {
            state.addNamedImport(
              property.value.value,
              property.key.value.toString(),
            );
          }
        }
      }
      break;
  }
}
