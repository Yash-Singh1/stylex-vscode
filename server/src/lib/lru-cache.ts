// Based off https://github.com/trekhleb/javascript-algorithms/blob/master/src/data-structures/lru-cache/LRUCacheOnMap.js

class LRUCache<K, V> extends Map<K, V> {
  private capacity: number;

  constructor(capacity: number) {
    super();
    this.capacity = capacity;
  }

  get(key: K) {
    if (!super.has(key)) return undefined;
    const val = super.get(key)!;
    super.delete(key);
    super.set(key, val);
    return val;
  }

  set(key: K, val: V) {
    super.delete(key);
    super.set(key, val);

    if (super.size > this.capacity) {
      super.delete(super.keys().return!().value as K);
    }

    return this;
  }
}

export default LRUCache;
