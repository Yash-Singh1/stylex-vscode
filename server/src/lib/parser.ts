import type { Module } from "@swc/types";
import type { CancellationToken } from "vscode-languageserver";
import type { TextDocument } from "vscode-languageserver-textdocument";

type Parser = typeof import("../../node_modules/@swc/wasm-web");

const PARSE_CANCELLED_MESSAGE = "[INFO] Parse Job Cancelled";

export function parse({
  source,
  languageId,
  parser,
  token,
}: {
  source: string;
  languageId: string;
  parser: Parser;
  token: CancellationToken;
}) {
  return new Promise<Module>((resolve, reject) => {
    token.onCancellationRequested(() => {
      reject(PARSE_CANCELLED_MESSAGE);
    });

    if (token.isCancellationRequested) {
      reject(PARSE_CANCELLED_MESSAGE);
    }

    parser
      .parse(source, {
        syntax: "typescript",
        tsx: languageId.endsWith("react"),
        target: "es2022",
        comments: true,
        decorators: true,
        dynamicImport: true,
      })
      .then((value) => resolve(value))
      .catch(reject);
  });
}

export function calculateStartOffset(textDocument: {
  getText(range: {
    start: { line: number; character: number };
    end: { line: number; character: number };
  }): string;
}) {
  let startOffset = 0;
  let line = textDocument
    .getText({
      start: { line: 0, character: 0 },
      end: { line: 1, character: 0 },
    })
    .slice(0, -1);
  let currentLine = 1;
  let multilineComment = false;

  while (
    !line.trim() ||
    line.trim().startsWith("//") ||
    line.trim().startsWith("/*") ||
    multilineComment
  ) {
    let changes = true;
    while (changes) {
      changes = false;
      if (!multilineComment && line.trim().startsWith("/*")) {
        multilineComment = true;
        startOffset += line.indexOf("/*") + 2;
        line = line.trim().slice(2);
        changes = true;
      }

      const multilineCommentEnd = line.indexOf("*/");
      if (multilineComment && multilineCommentEnd !== -1) {
        startOffset += multilineCommentEnd + 2;
        multilineComment = false;
        line = line.slice(multilineCommentEnd + 2);
        changes = true;
      }
    }

    if (!line.trim() || line.trim().startsWith("//") || multilineComment) {
      startOffset += line.length + 1;
      line = textDocument
        .getText({
          start: { line: currentLine, character: 0 },
          end: { line: currentLine + 1, character: 0 },
        })
        .slice(0, -1);
      ++currentLine;
    } else {
      break;
    }
  }

  startOffset += /^\s*/.exec(line)![0].length;

  return startOffset;
}
