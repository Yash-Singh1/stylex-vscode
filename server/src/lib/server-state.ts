import type { Module } from "@swc/types";
import type { ColorInformation } from "vscode-languageserver";
import type { StringAsBytes } from "./string-bytes";
import { CSSVirtualDocument } from "./virtual-document";
import {
  getCSSLanguageService,
  type LanguageService as CSSLanguageService,
} from "vscode-css-languageservice";

interface IServerState {
  setupCSSLanguageService(): void;
}

export default class ServerState implements IServerState {
  public static readonly STYLEX_CUSTOM_PROPERTY = "stylex-lsp-custom-property";

  public parserCache = new Map<string, Module>();
  public colorCache = new Map<string, ColorInformation[]>();
  public bytePrefixCache = new Map<string, StringAsBytes>();
  public virtualDocumentFactory = new CSSVirtualDocument();
  public cssLanguageService: CSSLanguageService | null = null;

  private cssLanguageServiceIsSetup = false;

  constructor() {}

  public setupCSSLanguageService() {
    if (this.cssLanguageServiceIsSetup) return;

    this.cssLanguageService = getCSSLanguageService();
    this.cssLanguageService.configure({
      completion: {
        completePropertyWithSemicolon: false,
        triggerPropertyValueCompletion: false,
      },
    });
    this.cssLanguageService.setDataProviders(true, [
      {
        provideAtDirectives() {
          return [];
        },
        providePseudoClasses() {
          return [];
        },
        providePseudoElements() {
          return [];
        },
        provideProperties() {
          return [
            {
              name: ServerState.STYLEX_CUSTOM_PROPERTY,
              restrictions: [
                "enum",
                "time",
                "timing-function",
                "box",
                "color",
                "repeat",
                "url",
                "line-style",
                "image",
                "length",
                "identifier",
                "number(0-1)",
                "number",
                "font",
                "string",
                "angle",
                "integer",
                "property",
                "percentage",
                "unicode-range",
                "line-width",
                "geometry-box",
                "position",
                "positon",
                "shape",
              ],
            },
          ];
        },
      },
    ]);
    this.cssLanguageServiceIsSetup = true;
  }
}
