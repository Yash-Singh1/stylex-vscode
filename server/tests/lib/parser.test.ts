import { test, expect, describe } from "vitest";

import { calculateStartOffset, parse } from "../../src/lib/parser";
import { TextDocument } from "vscode-languageserver-textdocument";
import { testParser } from "../helpers/parseSetup";

describe("parse", () => {
  testParser("parses a simple file", async ({ parser }) => {
    const source = `import stylex from "@stylexjs/stylex";

const styles = stylex.create({
  red: {
    color: "red",
  },
});
`;

    const module = await parse({
      source,
      languageId: "typescript",
      parser,
      token: {
        isCancellationRequested: false,
        onCancellationRequested: () => {
          return { dispose() {} };
        },
      },
    });

    expect(module).toMatchSnapshot();
  });

  testParser(
    "handles typescript that conflicts with tsx",
    async ({ parser }) => {
      const source = `import stylex from "@stylexjs/stylex";

interface Main {
  foo: string;
}

const config = <Main>{}; // This should not be parsed as a JSX element
`;

      let module = await parse({
        source,
        languageId: "typescript",
        parser,
        token: {
          isCancellationRequested: false,
          onCancellationRequested: () => {
            return { dispose() {} };
          },
        },
      });

      expect(module).toMatchSnapshot();

      const source2 = `import stylex from "@stylexjs/stylex";

      interface Main {
        foo: string;
      }
      
      const config = <Main>{"hello world"}</Main>; // This should be parsed as a JSX element
      `;
      module = await parse({
        source: source2,
        languageId: "typescriptreact",
        parser,
        token: {
          isCancellationRequested: false,
          onCancellationRequested: () => {
            return { dispose() {} };
          },
        },
      });

      expect(module).toMatchSnapshot();
    },
  );
});

describe("calculateStartOffset", () => {
  test("handles calculations correctly", () => {
    function realOffset(src: string) {
      return Math.max(src.indexOf("import"), src.indexOf("{"));
    }

    function assertOffsetCorrect(src: string) {
      expect(
        calculateStartOffset(
          TextDocument.create("untitled://example.ts", "typescript", 1, src),
        ),
        "Failed while calculating offset for " + src,
      ).toBe(realOffset(src));
    }

    assertOffsetCorrect(`import stylex from "@stylexjs/stylex"`);
    assertOffsetCorrect(`  import stylex from "@stylexjs/stylex"`);
    assertOffsetCorrect(`
    import stylex from "@stylexjs/stylex"`);

    assertOffsetCorrect(`import stylex from "@stylexjs/stylex"`);
    assertOffsetCorrect(`/**/
import stylex from "@stylexjs/stylex"`);
    assertOffsetCorrect(`
      import stylex from "@stylexjs/stylex"`);

    assertOffsetCorrect(`import stylex from "@stylexjs/stylex"`);
    assertOffsetCorrect(`/**
    @flow strict
    **/
   // more information

   /**/

   // help info

    
   import stylex from "@stylexjs/stylex"`);
  });
});
