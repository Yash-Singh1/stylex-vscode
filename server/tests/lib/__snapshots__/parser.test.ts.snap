// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parse > handles typescript that conflicts with tsx 1`] = `
{
  "body": [
    {
      "source": {
        "raw": ""@stylexjs/stylex"",
        "span": {
          "ctxt": 0,
          "end": 38,
          "start": 20,
        },
        "type": "StringLiteral",
        "value": "@stylexjs/stylex",
      },
      "span": {
        "ctxt": 0,
        "end": 39,
        "start": 1,
      },
      "specifiers": [
        {
          "local": {
            "optional": false,
            "span": {
              "ctxt": 2,
              "end": 14,
              "start": 8,
            },
            "type": "Identifier",
            "value": "stylex",
          },
          "span": {
            "ctxt": 0,
            "end": 14,
            "start": 8,
          },
          "type": "ImportDefaultSpecifier",
        },
      ],
      "type": "ImportDeclaration",
      "typeOnly": false,
      "with": null,
    },
    {
      "body": {
        "body": [
          {
            "computed": false,
            "init": null,
            "key": {
              "optional": false,
              "span": {
                "ctxt": 0,
                "end": 63,
                "start": 60,
              },
              "type": "Identifier",
              "value": "foo",
            },
            "optional": false,
            "params": [],
            "readonly": false,
            "span": {
              "ctxt": 0,
              "end": 72,
              "start": 60,
            },
            "type": "TsPropertySignature",
            "typeAnnotation": {
              "span": {
                "ctxt": 0,
                "end": 71,
                "start": 63,
              },
              "type": "TsTypeAnnotation",
              "typeAnnotation": {
                "kind": "string",
                "span": {
                  "ctxt": 0,
                  "end": 71,
                  "start": 65,
                },
                "type": "TsKeywordType",
              },
            },
            "typeParams": null,
          },
        ],
        "span": {
          "ctxt": 0,
          "end": 74,
          "start": 56,
        },
        "type": "TsInterfaceBody",
      },
      "declare": false,
      "extends": [],
      "id": {
        "optional": false,
        "span": {
          "ctxt": 2,
          "end": 55,
          "start": 51,
        },
        "type": "Identifier",
        "value": "Main",
      },
      "span": {
        "ctxt": 0,
        "end": 74,
        "start": 41,
      },
      "type": "TsInterfaceDeclaration",
      "typeParams": null,
    },
    {
      "declarations": [
        {
          "definite": false,
          "id": {
            "optional": false,
            "span": {
              "ctxt": 2,
              "end": 88,
              "start": 82,
            },
            "type": "Identifier",
            "typeAnnotation": null,
            "value": "config",
          },
          "init": {
            "expression": {
              "properties": [],
              "span": {
                "ctxt": 0,
                "end": 99,
                "start": 97,
              },
              "type": "ObjectExpression",
            },
            "span": {
              "ctxt": 0,
              "end": 99,
              "start": 91,
            },
            "type": "TsTypeAssertion",
            "typeAnnotation": {
              "span": {
                "ctxt": 0,
                "end": 96,
                "start": 92,
              },
              "type": "TsTypeReference",
              "typeName": {
                "optional": false,
                "span": {
                  "ctxt": 2,
                  "end": 96,
                  "start": 92,
                },
                "type": "Identifier",
                "value": "Main",
              },
              "typeParams": null,
            },
          },
          "span": {
            "ctxt": 0,
            "end": 99,
            "start": 82,
          },
          "type": "VariableDeclarator",
        },
      ],
      "declare": false,
      "kind": "const",
      "span": {
        "ctxt": 0,
        "end": 100,
        "start": 76,
      },
      "type": "VariableDeclaration",
    },
  ],
  "interpreter": null,
  "span": {
    "ctxt": 0,
    "end": 100,
    "start": 1,
  },
  "type": "Module",
}
`;

exports[`parse > handles typescript that conflicts with tsx 2`] = `
{
  "body": [
    {
      "source": {
        "raw": ""@stylexjs/stylex"",
        "span": {
          "ctxt": 0,
          "end": 185,
          "start": 167,
        },
        "type": "StringLiteral",
        "value": "@stylexjs/stylex",
      },
      "span": {
        "ctxt": 0,
        "end": 186,
        "start": 148,
      },
      "specifiers": [
        {
          "local": {
            "optional": false,
            "span": {
              "ctxt": 2,
              "end": 161,
              "start": 155,
            },
            "type": "Identifier",
            "value": "stylex",
          },
          "span": {
            "ctxt": 0,
            "end": 161,
            "start": 155,
          },
          "type": "ImportDefaultSpecifier",
        },
      ],
      "type": "ImportDeclaration",
      "typeOnly": false,
      "with": null,
    },
    {
      "body": {
        "body": [
          {
            "computed": false,
            "init": null,
            "key": {
              "optional": false,
              "span": {
                "ctxt": 0,
                "end": 222,
                "start": 219,
              },
              "type": "Identifier",
              "value": "foo",
            },
            "optional": false,
            "params": [],
            "readonly": false,
            "span": {
              "ctxt": 0,
              "end": 231,
              "start": 219,
            },
            "type": "TsPropertySignature",
            "typeAnnotation": {
              "span": {
                "ctxt": 0,
                "end": 230,
                "start": 222,
              },
              "type": "TsTypeAnnotation",
              "typeAnnotation": {
                "kind": "string",
                "span": {
                  "ctxt": 0,
                  "end": 230,
                  "start": 224,
                },
                "type": "TsKeywordType",
              },
            },
            "typeParams": null,
          },
        ],
        "span": {
          "ctxt": 0,
          "end": 239,
          "start": 209,
        },
        "type": "TsInterfaceBody",
      },
      "declare": false,
      "extends": [],
      "id": {
        "optional": false,
        "span": {
          "ctxt": 2,
          "end": 208,
          "start": 204,
        },
        "type": "Identifier",
        "value": "Main",
      },
      "span": {
        "ctxt": 0,
        "end": 239,
        "start": 194,
      },
      "type": "TsInterfaceDeclaration",
      "typeParams": null,
    },
    {
      "declarations": [
        {
          "definite": false,
          "id": {
            "optional": false,
            "span": {
              "ctxt": 2,
              "end": 265,
              "start": 259,
            },
            "type": "Identifier",
            "typeAnnotation": null,
            "value": "config",
          },
          "init": {
            "children": [
              {
                "expression": {
                  "raw": ""hello world"",
                  "span": {
                    "ctxt": 0,
                    "end": 288,
                    "start": 275,
                  },
                  "type": "StringLiteral",
                  "value": "hello world",
                },
                "span": {
                  "ctxt": 0,
                  "end": 289,
                  "start": 274,
                },
                "type": "JSXExpressionContainer",
              },
            ],
            "closing": {
              "name": {
                "optional": false,
                "span": {
                  "ctxt": 1,
                  "end": 295,
                  "start": 291,
                },
                "type": "Identifier",
                "value": "Main",
              },
              "span": {
                "ctxt": 0,
                "end": 296,
                "start": 289,
              },
              "type": "JSXClosingElement",
            },
            "opening": {
              "attributes": [],
              "name": {
                "optional": false,
                "span": {
                  "ctxt": 1,
                  "end": 273,
                  "start": 269,
                },
                "type": "Identifier",
                "value": "Main",
              },
              "selfClosing": false,
              "span": {
                "ctxt": 0,
                "end": 274,
                "start": 268,
              },
              "type": "JSXOpeningElement",
              "typeArguments": null,
            },
            "span": {
              "ctxt": 0,
              "end": 296,
              "start": 268,
            },
            "type": "JSXElement",
          },
          "span": {
            "ctxt": 0,
            "end": 296,
            "start": 259,
          },
          "type": "VariableDeclarator",
        },
      ],
      "declare": false,
      "kind": "const",
      "span": {
        "ctxt": 0,
        "end": 297,
        "start": 253,
      },
      "type": "VariableDeclaration",
    },
  ],
  "interpreter": null,
  "span": {
    "ctxt": 0,
    "end": 297,
    "start": 148,
  },
  "type": "Module",
}
`;

exports[`parse > parses a simple file 1`] = `
{
  "body": [
    {
      "source": {
        "raw": ""@stylexjs/stylex"",
        "span": {
          "ctxt": 0,
          "end": 38,
          "start": 20,
        },
        "type": "StringLiteral",
        "value": "@stylexjs/stylex",
      },
      "span": {
        "ctxt": 0,
        "end": 39,
        "start": 1,
      },
      "specifiers": [
        {
          "local": {
            "optional": false,
            "span": {
              "ctxt": 2,
              "end": 14,
              "start": 8,
            },
            "type": "Identifier",
            "value": "stylex",
          },
          "span": {
            "ctxt": 0,
            "end": 14,
            "start": 8,
          },
          "type": "ImportDefaultSpecifier",
        },
      ],
      "type": "ImportDeclaration",
      "typeOnly": false,
      "with": null,
    },
    {
      "declarations": [
        {
          "definite": false,
          "id": {
            "optional": false,
            "span": {
              "ctxt": 2,
              "end": 53,
              "start": 47,
            },
            "type": "Identifier",
            "typeAnnotation": null,
            "value": "styles",
          },
          "init": {
            "arguments": [
              {
                "expression": {
                  "properties": [
                    {
                      "key": {
                        "optional": false,
                        "span": {
                          "ctxt": 0,
                          "end": 77,
                          "start": 74,
                        },
                        "type": "Identifier",
                        "value": "red",
                      },
                      "type": "KeyValueProperty",
                      "value": {
                        "properties": [
                          {
                            "key": {
                              "optional": false,
                              "span": {
                                "ctxt": 0,
                                "end": 90,
                                "start": 85,
                              },
                              "type": "Identifier",
                              "value": "color",
                            },
                            "type": "KeyValueProperty",
                            "value": {
                              "raw": ""red"",
                              "span": {
                                "ctxt": 0,
                                "end": 97,
                                "start": 92,
                              },
                              "type": "StringLiteral",
                              "value": "red",
                            },
                          },
                        ],
                        "span": {
                          "ctxt": 0,
                          "end": 102,
                          "start": 79,
                        },
                        "type": "ObjectExpression",
                      },
                    },
                  ],
                  "span": {
                    "ctxt": 0,
                    "end": 105,
                    "start": 70,
                  },
                  "type": "ObjectExpression",
                },
                "spread": null,
              },
            ],
            "callee": {
              "object": {
                "optional": false,
                "span": {
                  "ctxt": 2,
                  "end": 62,
                  "start": 56,
                },
                "type": "Identifier",
                "value": "stylex",
              },
              "property": {
                "optional": false,
                "span": {
                  "ctxt": 0,
                  "end": 69,
                  "start": 63,
                },
                "type": "Identifier",
                "value": "create",
              },
              "span": {
                "ctxt": 0,
                "end": 69,
                "start": 56,
              },
              "type": "MemberExpression",
            },
            "span": {
              "ctxt": 0,
              "end": 106,
              "start": 56,
            },
            "type": "CallExpression",
            "typeArguments": null,
          },
          "span": {
            "ctxt": 0,
            "end": 106,
            "start": 47,
          },
          "type": "VariableDeclarator",
        },
      ],
      "declare": false,
      "kind": "const",
      "span": {
        "ctxt": 0,
        "end": 107,
        "start": 41,
      },
      "type": "VariableDeclaration",
    },
  ],
  "interpreter": null,
  "span": {
    "ctxt": 0,
    "end": 107,
    "start": 1,
  },
  "type": "Module",
}
`;
